x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/DetectionRule.java

```diff
@@ -55,87 +57,63 @@
+                            scope.put(entry.getKey(), filterRefParser.parse(scopeParser, null));
+                        }
+                    }
                 }
-                throw new IllegalArgumentException("Unsupported token [" + p.currentToken() + "]");
-            }, CONDITIONS_CONNECTIVE_FIELD, ValueType.STRING);
+                return new RuleScope(scope);
+            }, SCOPE_FIELD);
             parser.declareObjectArray(Builder::setConditions, (p, c) ->
                     RuleCondition.PARSERS.get(parserType).apply(p, c), CONDITIONS_FIELD);
         }
     }
 
     private final EnumSet<RuleAction> actions;
-    private final String targetFieldName;
-    private final String targetFieldValue;
-    private final Connective conditionsConnective;
+    private final RuleScope scope;
     private final List<RuleCondition> conditions;
 
-    private DetectionRule(EnumSet<RuleAction> actions, @Nullable String targetFieldName, @Nullable String targetFieldValue,
-                          Connective conditionsConnective, List<RuleCondition> conditions) {
+    private DetectionRule(EnumSet<RuleAction> actions, RuleScope scope, List<RuleCondition> conditions) {
         this.actions = Objects.requireNonNull(actions);
-        this.targetFieldName = targetFieldName;
-        this.targetFieldValue = targetFieldValue;
-        this.conditionsConnective = Objects.requireNonNull(conditionsConnective);
+        this.scope = Objects.requireNonNull(scope);
         this.conditions = Collections.unmodifiableList(conditions);
     }
 
     public DetectionRule(StreamInput in) throws IOException {
-        actions = EnumSet.noneOf(RuleAction.class);
-        if (in.getVersion().before(Version.V_6_2_0)) {
-            actions.add(RuleAction.readFromStream(in));
-        } else {
-            int actionsCount = in.readVInt();
-            for (int i = 0; i < actionsCount; ++i) {
-                actions.add(RuleAction.readFromStream(in));
-            }
-        }
-
-        conditionsConnective = Connective.readFromStream(in);
-        int size = in.readVInt();
-        conditions = new ArrayList<>(size);
-        for (int i = 0; i < size; i++) {
-            conditions.add(new RuleCondition(in));
-        }
-        targetFieldName = in.readOptionalString();
-        targetFieldValue = in.readOptionalString();
+        actions = in.readEnumSet(RuleAction.class);
+        scope = new RuleScope(in);
+        conditions = in.readList(RuleCondition::new);
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        if (out.getVersion().before(Version.V_6_2_0)) {
-            // Only filter_results is supported prior to 6.2.0
-            RuleAction.FILTER_RESULTS.writeTo(out);
-        } else {
-            out.writeVInt(actions.size());
-            for (RuleAction action : actions) {
-                action.writeTo(out);
-            }
+        if (out.getVersion().onOrAfter(Version.V_6_4_0)) {
```

Comment:
OK, on the input side we're relying on the rules list in 6.3 and below being an empty list.  We discussed that and said it was fine because nobody should have used a rule before 6.4, and if they did it was unsupported and hence acceptable for the transport to break.

On the output side I think a change is necessary though.  The way it is at the moment if you have a list of 2 rules in 6.4 then that will be serialised as a list of 2 objects each with no fields.  Instead I think it needs to be serialised as an empty list, because 6.3 won't understand the objects with no fields.  (Let me know if that's wrong.)

If my thinking is correct then this version check can be removed but instead we need on in `Detector.java` around this line:

```
out.writeList(rules);
```

If the remote node is below version 6.4 it needs to write an empty list instead of `rules`.

Reply:
You're absolutely right. Will push a fix for this shortly.